src/cell_group.hpp

```diff
@@ -143,51 +143,61 @@ class cell_group {
     }
 
     void advance(time_type tfinal, time_type dt) {
-        while (lowered_.time()<tfinal) {
-            // take any pending samples
-            time_type cell_time = lowered_.time();
+        EXPECTS(lowered_.state_synchronized());
```

Comment:
This function is getting very long: Would it be smart to split it in multiple steps to improve readability?